File: 287-find-the-duplicate-number/find-the-duplicate-number.py
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        slow = 0
        fast = 0

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]

            if slow == fast:
                break

        slow1 = 0
        while True:
            slow1 = nums[slow1]
            slow = nums[slow]

            if slow == slow1:
                break

        return slow

File: 287-find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        // floyd's cycle detection - tortoise and hare (in arrays)
        int slow = nums[0];
        int fast = nums[nums[0]];

        while (slow != fast){
            slow = nums[slow];
            fast = nums[nums[fast]];
        }

        fast = 0;
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return fast;
    }
};